m3/src/main/java/com/uber/m3/tally/m3/M3Reporter.java

```diff
@@ -0,0 +1,831 @@
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.Phaser;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import org.apache.thrift.transport.TTransportException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * An M3 implementation of a {@link StatsReporter}.
+ */
+public class M3Reporter implements StatsReporter, AutoCloseable {
+    public static final String SERVICE_TAG = "service";
+    public static final String ENV_TAG = "env";
+    public static final String HOST_TAG = "host";
+    public static final String DEFAULT_TAG_VALUE = "default";
+
+    public static final String DEFAULT_HISTOGRAM_BUCKET_ID_NAME = "bucketid";
+    public static final String DEFAULT_HISTOGRAM_BUCKET_NAME = "bucket";
+    public static final int DEFAULT_HISTOGRAM_BUCKET_TAG_PRECISION = 6;
+
+    private static final Logger LOG = LoggerFactory.getLogger(M3Reporter.class);
+
+    private static final int MAX_PROCESSOR_WAIT_ON_CLOSE_MILLIS = 1000;
+
+    private static final int DEFAULT_FLUSH_TRY_COUNT = 2;
+    private static final int FLUSH_RETRY_DELAY_MILLIS = 200;
+
+    private static final int DEFAULT_METRIC_SIZE = 100;
+
+    private static final int DEFAULT_MAX_QUEUE_SIZE = 4096;
+    private static final int DEFAULT_MAX_PACKET_SIZE = 1440;
+
+    private static final int THREAD_POOL_SIZE = 10;
+
+    private static final int EMIT_METRIC_BATCH_OVERHEAD = 19;
+    private static final int MIN_METRIC_BUCKET_ID_TAG_LENGTH = 4;
+
+    private M3.Client client;
+
+    private final MetricBatch metricBatch = new MetricBatch();
+
+    private final Object calcLock = new Object();
+    private TCalcTransport calc;
+    private TProtocol calcProtocol;
+
+    private int freeBytes;
+
+    private String bucketIdTagName;
+    private String bucketTagName;
+    private String bucketValFmt;
+
+    // Holds metrics to be flushed
+    private final BlockingQueue<SizedMetric> metricQueue;
+    // The service used to execute metric flushes
+    private ExecutorService executor = Executors.newFixedThreadPool(THREAD_POOL_SIZE);
+    // Used to keep track of how many processors are in progress in
+    // order to wait for them to finish before closing
+    private Phaser phaser = new Phaser(1);
+
+    private final Object isOpenLock = new Object();
+    private boolean isOpen = true;
+
+    private TTransport transport;
+
+    // Use inner Builder class to construct an M3Reporter
+    private M3Reporter(Builder builder) {
+        try {
+            // Builder verifies non-null, non-empty socketAddresses
+            SocketAddress[] socketAddresses = builder.socketAddresses;
+
+            TProtocolFactory protocolFactory = new TCompactProtocol.Factory();
+
+            if (socketAddresses.length > 1) {
+                throw new UnsupportedOperationException(
+                    "Support for a multiple SocketAddress transport not yet implemented."
+                );
+            } else {
```

Comment:
don't need the else here, you're throwing an exception in the if condition.

Reply:
Good point. I have since implemented support for multiple addresses.